<h1 id="about-the-cttv-target-validation-platform">About the CTTV target validation platform</h1>
<p>The Center for Therapeutic Target Validation (CTTV) platform brings together evidence which associates potential
drug targets with diseases. Its aim is to help users identify and prioritise potential targets for further investigation.</p>
<p>The platform brings together evidence from multiple data types that can be relevant to target identification and prioritisation.
These data are described in detail on the <a href="/data_sources">data sources</a> page.</p>
<p>A target can be a protein, protein complex or RNA molecule. We integrate evidence through the gene that codes for
the target using <a href="http://www.ensembl.org/info/genome/genebuild/genome_annotation.html">Ensembl gene identifiers</a>. We display the
target by the <a href="http://www.genenames.org">HGNC gene symbol</a> but you can search for it using protein or gene names
and most synonyms as explained in <a href="/faq#what_can_I search_for">What can I search for</a>.</p>
<p>We map diseases to terms in the <a href="http://www.ebi.ac.uk/efo/">Experimental Factor Ontology</a> (EFO), which enables us to
describe relationships between diseases. For more details see <a href="/faq#diseases">How are diseases described?</a></p>
<h2 id="getting-started">Getting Started</h2>
<p>The platform supports workflows starting from either a target or disease and presents the evidence for target – disease
associations in a number of ways through Association and Evidence pages. We also present related information on the target
or the disease itself through the <a href="#target_profile">Target Profile</a> and <a href="#disease_profile">Disease Profile</a> pages.</p>
<p>A search box is available on the home page or at the top of every other page. To find a target of interest, either enter
the gene or protein name, symbol or synonym. To find a disease you can search for the common disease name or its synonyms
(including some abbreviations) provided the disease label and its synonyms are part of the <a href="http://www.ebi.ac.uk/efo/">EFO</a>.</p>
<h2 id="target-association-page">Target Association page</h2>
<p>After searching for a target and selecting the target you will be taken to the Target Association page which provides an
overview of all diseases associated to a target.</p>
<p>Three views of the associations are available organised by the tabs at the top:</p>
<ul>
<li>Bubbles view,</li>
<li>Table view and</li>
<li>Tree view.</li>
</ul>
<p>Filters are available on the left of the page to restrict the associations that are displayed by data type or to focus
on a therapeutic area (see <a href="#filters">filters</a>).</p>
<h3 id="bubbles-view">Bubbles View</h3>
<p>This view groups diseases into &#39;bubbles&#39; by therapeutic area. Each large bubble corresponds to a therapeutic area and
consists of smaller bubbles representing diseases within this area. A disease can belong to several therapeutic areas
and therefore can appear within more than one therapeutic area bubble.</p>
<p>The strength of the association between the target and a particular disease is represented by the size of the corresponding
disease bubble and the depth of its blue colour: The larger the bubble and the deeper the colour, the stronger the association.</p>
<p>Mouse over a bubble to display the full name of the disease and the score associating it with the target. Clicking on a
bubble opens a pop-up with several links: Click on the flower icon or to ‘View evidence details’ to view the detailed
<a href="#evidence">Target-Disease Evidence</a> page for the association, click on the ‘P’ icon to view the Disease Profile page
or the ‘A’ icon to view all the target associations of that disease.</p>
<p>To zoom in on a particular therapeutic area, click on the corresponding therapeutic area filter on the left hand side.
To zoom out of a particular therapeutic area, click on the therapeutic area filter again. You can also filter the bubbles
by data types using the corresponding filter on the left hand side.</p>
<h3 id="table-view">Table View</h3>
<p>This view lists all diseases associated to a particular gene ordered by the association score. Scores are represented by the depth
of blue colour. Cells representing associations with no evidence are coloured white. In this view we show evidence from the
lower parts of the disease hierarchy aggregated into higher terms. So for instance in the <a href="https://www.targetvalidation.org/target/ENSG00000157764/associations">BRAF Target Association</a>
page, evidence is shown for higher level terms such as skin disease, or carcinoma.</p>
<p>The filters work as before on this view and in addition the associations can be ordered by the association scores for individual data
types by clicking the arrow in the table header. To order by several data types, shift-click the table headers. There is also a search box
available and typing into the box will restrict the diseases displayed to those that match the search string. For example try typing &quot;Mela&quot;
into the search box on the <a href="https://www.targetvalidation.org/target/ENSG00000157764/associations">BRAF</a> table view to restrict the diseases
to types of melanoma.</p>
<h3 id="tree-view">Tree View</h3>
<p>The ‘Tree’ view allows visualisation of evidence across the therapeutic areas in a tree representing the relationships of diseases.
Therapeutic areas have a square symbol, while the other disease nodes are circles. The depth of the blue fill of the node indicates the
association score at that node. Top-level disease categories are shown at a fixed position. Clicking on a disease opens a pop-up menu
that provides links to various other pages, but also allows branches of the tree to be collapsed down to simplify the visualisation.
Clicking on a collapsed disease will allow the branches to be opened again.</p>
<h2 id="disease-association-page">Disease Association page</h2>
<p>This page provides a table view of all potential targets associated to a disease ordered by the association score. The depth of blue colour
for the cells in the association score and for the individual data types again indicates the strength of the association. Targets can be filtered
by data type and also by involvement in Reactome reaction pathways as described <a href="#filters">below</a>. The associations can be ordered
by the association scores for individual data types by clicking the arrow in the table header. To order by several data types, shift-click
the table headers.</p>
<p>There is also a search box available and typing into the box will restrict the targets displayed to those that match the search string.
For example try typing &quot;IL&quot; into the search box on the <a href="https://www.targetvalidation.org/disease/EFO_0000270/associations">Asthma</a> table
view to restrict the targets largely to Interleukins.</p>
<h2 id="-a-name-evidence-a-target-disease-evidence-page"><a name="evidence"></a>Target-Disease Evidence page</h2>
<p>After viewing an association in either the Target Association page or Disease Association page, details of the underlying evidence
can be accessed by clicking on the target/disease or the evidence cells of the table view, or by selecting the evidence flower, or
&#39;View evidence details&#39; in the bubble or tree views. This displays the Target-Disease Evidence page.  At the top of this page is a panel
indicating the association that is being examined, and a flower that shows the score for each data type as the petals of a flower.  The
extent of the blue fill of the flower petal indicates the strength of the association for that data type. Details of the data types can be
found <a href="/data_sources">here</a>.</p>
<p>The sections below contain the details of the evidence available for this association in each data type. If evidence is available
for a particularly datatype the section header will contain bold black text and clicking on the header will open our a table that describes the
available data and the source. In some cases (e.g. Genetic associations) there will also be a tab available that displays a graphical
view of the data (e.g. through a genome browser). If there is no evidence in a datatype the header will be greyed out and clicking on
the header will open out a section indicating there is &quot;No data available&quot;.</p>
<p>In each of the evidence tables, the columns may be sorted by the data, by clicking on the arrows in the table header. The tables can also be
searched for matches to text by typing in the &#39;Search:&#39; box. Details of the content of the tables follow:</p>
<h3 id="genetic-associations">Genetic associations</h3>
<p>This section splits into common disease evidence from Genome Wide Association Studies (<a href="/data_sources#gwas">GWAS catalog</a>),
and rare disease evidence from Mendelian disease (<a href="/data_sources#uniprot">UniProt</a> and <a href="/data_sources#eva">European Variation Archive</a>).</p>
<p>For GWAS data we provide the identity of the SNP that was reported as the strongest association withe reported p value and the evidence that
links the SNP to the target.</p>
<p>For rare disease, we report the mutation if it is available and the curated cosequence of the mutation.</p>
<h3 id="somatic-mutations">Somatic mutations</h3>
<p>Somatic mutation evidence is provided primarily for cancers, from the <a href="/data_sources#census">Cancer Gene Census</a>. This data is a summary of multiple
and sometimes many somatic mutations that support the association of a target to disease, and we provide the details of the mutation types
observed that contribute to the overall summary annotation.</p>
<p>In addition somatic mutation data can also come from the <a href="/data_sources#eva_somatic">European Variation Archive</a></p>
<h3 id="drugs">Drugs</h3>
<p>This section lists FDA approved and marketed drugs that associate a specific target to a specific disease. The drug to target association
and drug to disease associations are curated by <a href="/data_sources#chembl">ChEMBL</a> from multiple evidence sources as listed.</p>
<h3 id="affected-pathways">Affected pathways</h3>
<p>Evidence in this section reflects specific curation in <a href="/data_sources#reactome">Reactome</a> of a molecular mechanism that affects a pathway leading to disease.</p>
<h3 id="rna-expression">RNA expression</h3>
<p>The RNA expression data table shows gene changes for which the log2 fold change is greater than 1 or less than -1 and p-value is less than 0.05
(after Benjamini &amp; Hochberg (1995) FDR correction). The percentile rank column shows the relative rank of the gene expression fold change of
that gene in comparison to all other genes for this experiment.</p>
<h3 id="text-mining">Text mining</h3>
<p>The Literature Mining data table shows the most relevant research articles based on the number of times an association between a gene and a
disease is found in sentences across the article and highlights the terms in the relevant sentences.  Clicking on the number in the Matched
sentences column will reveal a pop-up with all relevant sentences from the article.</p>
<h3 id="animal-models">Animal models</h3>
<p>The table shows the mapping of the human and mouse phenotypes together withthe identity of the mouse knockout mutation that provides the
evidence from the <a href="/data_sources#mouse">Phenodigm</a> database.</p>
<h2 id="-a-name-target_profile-a-target-profile-page"><a name="target_profile"></a>Target Profile Page</h2>
<p>The Target Profile page summarizes relevant information about a particular target outside of the context of a specific disease association.
The Target profile can be accessed either by clicking the target name in the Target-Disease Evidence page, or the &quot;View BRAF profile&quot; link
in the Target Association page.</p>
<p>The sections of the page open out when they are clicked to reveal the information either as text or tables, or as a graphical display or both.
Information available here includes:</p>
<ul>
<li>A summary of the target function from <a href="http://www.uniprot.org">UniProt</a>,</li>
<li>Protein and gene synonyms.</li>
<li>Protein information from <a href="http://www.uniprot.org">UniProt</a> including a graphical protein feature viewer,</li>
<li>Variants, isoforms and genomic context information section displaying common disease variants and rare disease mutations associated with diseases
relative to the alternative transcripts (isoforms) of the gene, supplied from <a href="http://www.ensembl.org">Ensembl</a>,</li>
<li>Protein Baseline Expression displays data from the <a href="http://www.proteinatlas.org/">Human Protein Atlas</a> expression analysis at either the protein or
RNA level. Expression levels are classified as low, medium and high as <a href="http://www.proteinatlas.org/about/assays+annotation">described</a>,</li>
<li>RNA Baseline Expression displays RNA expression levels from large scale RNA-seq profiling of normal tissues. We include data from six different
studies to provide a wide coverage of different tissues. Data is provided by the EMBL-EBI <a href="https://www.ebi.ac.uk/gxa/home">Expression Atlas</a>,</li>
<li>Gene Ontology terms,</li>
<li>Protein Structure information from <a href="https://www.ebi.ac.uk/pdbe">PDBe</a>,</li>
<li>Reaction pathway information from <a href="http://www.reactome.org">Reactome</a>,</li>
<li>Information on FDA approved and marketed drugs that interact with the target from the <a href="https://www.ebi.ac.uk/chembl/">ChEMBLl</a> database. In the
Target Profile page, all drugs for this target are displayed independent of disease,</li>
<li>Target Family</li>
<li>Bibliography information including the sources of literature data used to annotate the corresponding protein entry in
EMBL-EBI <a href="http://www.uniprot.org">UniProt</a> and is provided through <a href="https://europepmc.org/">EUROPE PubMed Central</a>.</li>
</ul>
<h2 id="-a-name-disease_profile-a-disease-profile-page"><a name="disease_profile"></a>Disease Profile page</h2>
<p>The Disease Profile page provides a summary of the disease and is accessed by clicking on the disease name in name in the Target-Disease Evidence page, or
the &quot;View disease profile&quot; link in the Disease Association page, or the ‘P’ icon  from a pop-up menu.</p>
<p>The Disease Profile page provides a description of the disease and its synonyms from the <a href="/faq#efo">EFO</a> and its relationship
to other diseases in the ontology. This page also displays the relationships of a disease to its immediate parents and children within the
<a href="/faq#efo">EFO</a> hierarchy. Clicking on ‘View target associations’ links to the Disease Association page for that disease.</p>
<p>The Drug section on this page provides a list of all FDA-approved and marketed drugs to treat this disease.</p>
<h2 id="-a-name-filters-a-filters"><a name="filters"></a>Filters</h2>
<p>A general concept in the platform is that a set of associations can be filtered according to properties of the associations.
The following filters are currently available:</p>
<ul>
<li>Data types: The diseases or targets displayed on the Association page are restricted to those having evidence from the selected data types.</li>
<li>Therapeutic areas: This filter restricts diseases shown on the Target Association page by those in the therapeutic areas selected.</li>
<li>Pathway types: This filter restricts the targets shown on the Disease Association page to those occurring in the selected <a href="http://www.reactome.org">Reactome</a> pathway types and individual pathways.</li>
</ul>
<p>When a Data types or Pathway types filter is applied the number of associations shown on the page is updated, and the number of associations
available in the other filters on the page is also updated. Multiple Data types or Pathway types filters can be applied at the same time.
Where multiple views are available as in the Target Association page, the filters apply across each of the views.</p>
<p>The filters are displayed by default at the top level of the classification. Where additional sub-divisions are available, as for the
Pathway types, clicking the arrow next to a particular Pathway type shows the individual pathways that this type consists of. Clicking
on the arrow for a Therapeutic area filter lists the individual diseases within this area.</p>
<h2 id="-a-name-citation-a-citation"><a name="citation"></a>Citation</h2>
<p>If you find the CTTV Target Validation Platform useful, please consider citing us as below:</p>
<p>The CTTV Target Validation Platform <a href="www.targetvalidation.org">www.targetvalidation.org</a></p>
<h2 id="-a-name-support-a-support-and-further-information"><a name="support"></a>Support and Further Information</h2>
<p>For support and feedback on the platform email <a href="&#x6d;&#97;&#105;&#108;&#116;&#111;&#x3a;&#115;&#x75;&#112;&#x70;&#x6f;&#114;&#x74;&#x40;&#x74;&#97;&#114;&#103;&#101;&#x74;&#x76;&#x61;&#x6c;&#x69;&#x64;&#x61;&#116;&#105;&#111;&#x6e;&#x2e;&#x6f;&#114;&#x67;">&#115;&#x75;&#112;&#x70;&#x6f;&#114;&#x74;&#x40;&#x74;&#97;&#114;&#103;&#101;&#x74;&#x76;&#x61;&#x6c;&#x69;&#x64;&#x61;&#116;&#105;&#111;&#x6e;&#x2e;&#x6f;&#114;&#x67;</a></p>
<p>If you want to be kept up to date on CTTV and platform developments, email <a href="&#x6d;&#97;&#105;&#x6c;&#x74;&#111;&#58;&#x69;&#110;&#102;&#x6f;&#64;&#x74;&#97;&#x72;&#103;&#x65;&#x74;&#118;&#x61;&#108;&#x69;&#x64;&#x61;&#x74;&#105;&#111;&#110;&#46;&#x6f;&#114;&#103;">&#x69;&#110;&#102;&#x6f;&#64;&#x74;&#97;&#x72;&#103;&#x65;&#x74;&#118;&#x61;&#108;&#x69;&#x64;&#x61;&#x74;&#105;&#111;&#110;&#46;&#x6f;&#114;&#103;</a></p>
