<div>
  <!-- Stack the columns on mobile by making one full-width and the other half-width -->
  <div class="row">
    <div class="col-sm-3">

        <div class="panel panel-default" style="margin-top:20px">
            <div class="panel-heading">
    	      <h4 class="panel-title">Quick Links</h4>
            </div>
            <div class="panel-body">
                <ul class="plain arrow-icon">
                    <li><a href="#search_box">What can I search for?</a></li>
                    <li><a href="#data-provenance">Where does the data come from?</a></li>
                    <li><a href="#association-score">How is the association score calculated?</a></li>
                    <li><a href="#association-score-filter">What is the association strength filter?</a></li>
                    <li><a href="#association-score-filter-datatype">Will the score change if I filter by datatype?</a></li>
                    <li><a href="#disease-citation">How are the articles shown in the Bibliography chosen?</a></li>
                    <li><a href="#security">Is the information I enter into the CTTV Beta platform confidential?</a></li>
                </ul>
            </div>
        </div>

    </div>
    <div class="col-sm-9 faqs">
    <h2><a name="about">Frequently asked questions</a></h2>

    <h3><a id="search-box" name="search_box">What can I search for?</a></h3>
		<p>To find a target you can search for the gene or protein name and symbol (including their synonyms) as annotated in major biological databases.</p>
		<p>To find a disease you can search for the common disease name and its synonyms (including some abbreviations) as described in the <a href="http://www.ebi.ac.uk/efo/" target="_blank">EFO</a> ontology</p>

    <h3><a id="data-provenance" name="data-provenance">Where does the data come from?</h3>
        <p>The data in the CTTV Core platform comes from a number of different sources: 
        <ul>
        <li>The <a href="http://www.ebi.ac.uk/fgpt/gwas/" target="_blank">GWAS Catalog&nbsp;<i class="fa fa-external-link"></i></a> provides evidence on the association of targets (genes) and diseases from Genome Wide Association Studies (GWAS).  The GWAS Catalog provides a list of the most associated variants (SNPs) by genomic region with the disease form each study.  The association is assigned the most likely gene by assessing the probable link between SNP and gene. SNPs are assigned to a gene by first considering any deleterious consequence within the genes coding region, then whether it is located within the gene's introns or regulatory regions and finally identifying the gene promoter to which it is nearest if there is no other data. </li>
		<li><a href="http://www.uniprot.org" target="_blank">UniProt&nbsp;<i class="fa fa-external-link"></i></a> or the <a href="http://www.ebi.ac.uk/eva/?Home" target="_blank">European Variation Archive (EVA)&nbsp;<i class="fa fa-external-link"></i></a> supply information on rare genetic diseases (Mendelian diseases). In this data type, evidence that a target (gene) is associated with a disease comes primarily from deleterious mutation within the protein coding region of the gene in diseased individuals and segregation within families. UniProt curates evidence from the literature and public databases including <a href="http://www.ncbi.nlm.nih.gov/omim" target="_blank">Online Mendelian Inheritance in Man (OMIM)</a> relevant to disease for UniProt protein entries.  EVA collects information on all human variation.  For target - disease evidence we take clinical relevant data from EVA where the variation is conseidered pathogenic, in particular from the <a href="http://www.ncbi.nlm.nih.gov/clinvar/" target="_blank">ClinVar</a> archive.</li>
	    <li>The <a href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/" target="_blank">COSMIC&nbsp;<i class="fa fa-external-link"></i></a> database provides evidence from somatic mutations on likely cancer drivers as curated in the Cancer Gene Census</li>
		<li><a href="https://www.ebi.ac.uk/chembl/" target="_blank">CHEMBL&nbsp;<i class="fa fa-external-link"></i></a> provides evidence from known drugs where the drug can be connected to a disease and a known target. Currently, we are displaying data on FDA-approved drugs (i.e. those which are approved for clinical trials in the USA). Clinical indication data from other countries is being reviewed and processed.</li>
		<li>The <a href="http://www.ebi.ac.uk/gxa/home" target="_blank">Expression Atlas&nbsp;<i class="fa fa-external-link"></i></a> supplies evidence gene expression changes in studies that compare disease samples with normal samples, or in some cases with other disease samples. In addition the Expression Atlas provides baseline expression infomtation for each target, which can be found on the Target page.</li>
        <li><a target="_blank" href="http://europepmc.org/">Europe PubMed Central&nbsp;<i class="fa fa-external-link"></i></a> (Europe PMC for short) is a unique, free, information resource for biomedical and health researchers that provides access to 30.4 million+ abstracts and 3.3 million+ full text research articles from PubMed and PubMed Central. For CTTV, Europe PMC supplies evidence of links between genes and diseases by mining the titles, abstracts and full text research articles using an entity-to-entity recognition approach.</li>
	    <li>As a complement of Europe PMC text mined information, CTTV integrates also MEDLINE literature mined information from <a target="_blank" href="http://www.disgenet.org/">DisGeNET&nbsp;<i class="fa fa-external-link"></i></a> via a NLP-based approach.</li>
		<li><a href="http://www.reactome.org" target="_blank">Reactome&nbsp;<i class="fa fa-external-link"></i></a> identifies reaction pathways that are affected by pathogenic mutations.</li>
		<li><a href="http://www.sanger.ac.uk/resources/databases/phenodigm/" target="_blank">Phenodigm&nbsp;<i class="fa fa-external-link"></i></a> provides evidence on associations of targets and disease via knockout mouse models.  In this database the orthologous human gene to the one knocked out in the mouse is identified. the phenotypic effects in the mutant mouse are then mapped to phenotypes associated with human diseases and probable matches are identified and scored. </li>
        </ul>
        </p>

    <h3><a id="association-score" name="association-score">How is the association score calculated?</h3>
        <p>The overall association score between a target and a disease is a numerical value indicating the strength of the association.</p>
		<p>The value is in the range 0 to 1, with 1 indicating the strongest association.</p>
		<p>The association score is produced by combining the scores for the target and disease from the individual datatypes (genetic associations, known drugs, etc.), which themselves can combine several data sources. Each association from each data source provides a score that includes terms for the frequency that the association is seen compared to other associations, the strength or severity of the effect, and the confidence of the study that provided the association.  See detailed breakdown of data source scores(link) These scores also range between 0 and 1 and are summed after applying scaling factors to give datatype scores. 
			In this summing process the max score is capped at 1. Datatype scores are then summed again after scaling to give the overall association score.</p>

                <h3><a id="association-score-filter" name="association-score-filter">What is the association strength filter?</h3>
                <p>The score filter allows associations to be filtered by score. There are 4 filter settings that refine the associations shown from all associations at the minimum to only the strongest at the max setting. 
                    Applying the 2 highest filter setting also stretches the score range to provide higher resolutions at the highest scores.</p>
                    
                <h3><a id="association-score-filter-datatype" name="association-score-filter-datatype">Will the score change if I filter by datatype?</h3>
                <p>Filtering by data type now does not change the association score but instead requires that the associations have evidence form that data type. </p>
		
				<h3><a id="disease-citation" name="disease-citation">How are the articles shown in the Bibliography chosen?</h3>
				<p>The list of citations appearing in the Bibliography section of the target page comes from the the literature citations that are the sources of data used to annotate the corresponding protein entry in UniProt.</p>
			
			    <h3><a id="security" name="security">Is the information I enter into the CTTV Beta platform confidential?</h3>
			    <p>GSK, Sanger and the EBI are bound by a mutual confidentiality agreement. All information that you type in and retrieve from the Beta site is confidential and won't be disclosed to anyone outside these three organisations. </p>
                <p>However, if you have signed a confidentiality agreement with a third party about a particular project, please don't use the CTTV platform to retrieve information about this project.</p>
                <p>We use standard encryption methods to keep the site secure. </p>
                <p>We log information from user queries to improve the platform. The log data is stored anonymously, and is also kept confidential and used solely for the purposes of improving the platform. </p>
                <p>Emails sent between the domain names @gsk.com, @ebi.ac.uk, @sanger.ac.uk and @targetvalidation.org are encrypted and therefore secure.</p>
                <p>Any feedback that you provide about the site by email or in person is also kept confidential and won't be used for any purpose other than improving the platform.</p>
    </div>
  </div>

</div>