<!-- search app / results -->

<div>
	<h3 view-title>Frequently asked questions</h3>

	
	<div class="row">

		<!-- 

		<div class="col-md-2">
			
		</div>
 		--> 
		
		<div class="col-md-12 faqs">

			
			<div>
				<h4 id="search-box">What can I search for?</h4>
				<p>To find a target you can search for the gene or protein name and symbol (including their synonyms) as annotated in major biological databases.</p>
				<p>To find a disease you can search for the common disease name and its synonyms (including some abbreviations) as described in the <a href="http://www.ebi.ac.uk/efo/" target="_blank">EFO</a> ontology</p>

				<h4 id="data-provenance">Where does the data come from?</h4>
				<p>The data in the CTTV Core platform comes from a number of different sources: 
				    <ul>
				        <li>The <a href="http://www.ebi.ac.uk/fgpt/gwas/" target="_blank">GWAS Catalog</a> provides evidence on the association of targets (genes) and diseases from Genome Wide Association Studies (GWAS).  The GWAS Catalog provides a list of the most associated variants (SNPs) by genomic region with the disease form each study.  The association is assigned the most likely gene by assessing the probable link between SNP and gene. SNPs are assigned to a gene by first considering any deleterious consequence within the genes coding region, then whether it is located within the gene's introns or regulatory regions and finally identifying the gene promoter to which it is nearest if there is no other data. </li>
				        <li><a href="http://www.uniprot.org" target="_blank">UniProt</a> or the <a href="http://www.ebi.ac.uk/eva/?Home" target="_blank">European Variation Archive (EVA)</a> supply information on rare genetic diseases (Mendelian diseases). In this data type, evidence that a target (gene) is associated with a disease comes primarily from deleterious mutation within the protein coding region of the gene in diseased individuals and segregation within families. UniProt curates evidence from the literature and public databases including <a href="http://www.ncbi.nlm.nih.gov/omim" target="_blank">Online Mendelian Inheritance in Man (OMIM)</a> relevant to disease for UniProt protein entries.  EVA collects information on all human variation.  For target - disease evidence we take clinical relevant data from EVA where the variation is conseidered pathogenic, in particular from the <a href="http://www.ncbi.nlm.nih.gov/clinvar/" target="_blank">ClinVar</a> archive.</li>
				        <li>The <a href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/" target="_blank">COSMIC</a> database provides evidence from somatic mutations on likely cancer drivers as curated in the Cancer Gene Census</li>
				        <li><a href="https://www.ebi.ac.uk/chembl/" target="_blank">CHEMBL</a> provides evidence from known drugs where the drug can be connected to a disease and a known target. Currently we have data on marketed small molecule drugs.</li>
				        <li>The <a href="http://www.ebi.ac.uk/gxa/home" target="_blank">Expression Atlas</a> supplies evidence gene expression changes in studies that compare disease samples with normal samples, or in some cases with other disease samples. In addition the Expression Atlas provides baseline expression infomtation for each target, which can be found on the Target page.</li>
				        <li><a href="http://www.reactome.org" target="_blank">Reactome</a> identifies reaction pathways that are affected by pathogenic mutations.  </li>
				        <li><a href="http://www.sanger.ac.uk/resources/databases/phenodigm/" target="_blank">Phenodigm</a> provides evidence on associations of targets and disease via knockout mouse models.  In this database the orthologous human gene to the one knocked out in the mouse is identified. the phenotypic effects in the mutant mouse are then mapped to phenotypes associated with human diseases and probable matches are identified and scored. </li>
				    </ul>
				</p>

				<h4 id="association-score">How is the association score calculated?</h4>
				<p>The association score between a target and a disease is a numerical value indicating the confidence of their association.</p>
				<p>The value is in the range -1 to 1, where:
				    <ul>
				        <li> 1 - is the maximum confidence of an association between the gene and the disease</li>
				        <li> 0 - there is no known association between the gene and the disease</li>
				        <li>-1 - is the maximum confidence of a negative association between the gene and the disease</li>
				    </ul>
				</p>
				<p>The association score is a number resulting from the combination of scores provided by the datasources for the target and disease taken into consideration. The datasource scores range between -1 and 1, describing the confidence of the association according to the data provided by the datasource.<br />
				   Each association datatype (genetic associations, known drugs, etc.) results from the integration of one or multiple datasources. The association score for each datatype depends on the confidence in each association and also in the number of associations for a particular datasource.<br />
				   In the current implementation the association score is calculated as the maximum score available from any datasource. The datasource considered to calculate the association score can be filtered to customise the calculation.
				</p>
				<p>Please note than at the moment there are no negative associations listed, and the method to calculate the association and the datasource scores is temporary and will be revised in the future</p>

				<h4 id="disease-citation">How are the articles shown in the Bibliography chosen?</h4>
				<p>The list of citations appearing in the Bibliography section of the target page comes from the the literature citations that are the sources of data used to annotate the corresponding protein entry in UniProt.</p>
			</div>
			
		</div>

	</div>
	



	
</div>
<!-- end search results-->
