<view-title>Target Validation Platform Release Notes</view-title>

<h1>Target Validation Platform Release Notes</h1>

<div class="row">
    <!-- release 1.1 -->
	<div class="col-md-12">
		<div>
			<h2>Release 1.1 (release 2016-04-18)</h2>
			<h3>Status</h3>
			<p>This is the second public version of the Target Validation Platform</p>
			<p>The Target Validation Platform integrates over 2.1M target-disease associations covering 7,954 diseases and 24,774 targets from the following public resources:</p>
			<ul>
				<li>9,445 ontology classes from <a href="http://www.ebi.ac.uk/efo/">the Experimental Factor Ontology</a> (EFO version 2.69)</li>
				<li>10,571 genetic evidence from <a href="http://www.uniprot.org/">UniProt</a>, a comprehensive, high-quality and freely accessible resource of protein sequence and functional information</li>
				<li>19,857 genetic evidence from <a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a>, an open-access database of all types of genetic variation data from human</li>
				<li>25,536 genetic evidence from <a href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>, the NHGRI-EBI Catalog of published genome-wide association studies</li>
				<li>29,812 evidence from <a href="https://www.ebi.ac.uk/chembl/">ChEMBL</a>, a database that contains compound bioactivity data against drug targets</li>
				<li>433,809 evidence from <a href="https://www.ebi.ac.uk/gxa/home">Expression Atlas</a> that provides information on gene expression patterns under different biological conditions</li>
				<li>9,468 affected pathway evidence from <a href="http://www.reactome.org/">Reactome</a>, a free, open-source, curated and peer reviewed pathway database</li>
				<li>9,790 evidence from the <a href="http://cancer.sanger.ac.uk/census/">COSMIC cancer Gene Census</a>, a catalogue of genes for which mutations have been causally implicated in cancer</li>
				<li>457 somatic evidence from <a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a></li>
				<li>395,407 evidence from <a href="https://www.sanger.ac.uk/resources/databases/phenodigm">PhenoDigm</a>, a database of animal model matches ranked by their phenotypic similarity to human disorder</li>
				<li>3,166,744 scientific literature evidence from co-occurence mining in <a href="https://europepmc.org/">Europe PMC</a></li>
			</ul>
			<p>This release includes the following features described in details in the <a href="/about">documentation</a>:</p>
			<ul>
				<li>REST API: programmatic access to the Target Validation Platform (see <a href="/documentation/api">documentation</a>) </li>
				<li>Data dumps: direct access to the data (see <a href="/dumps/">downloads</a>)</li>
				<li>Look and Feel: new 'Open Targets' logo and new color palette</li>
				<li>Search results: results of a search are presented by order of relevance and indicates the number of diseases associated for any target and the number of targets associated for any disease. Matched terms are highlighted</li>
				<li>Target Association page: Therapeutic Areas are now presented as individual facets that can be combined. This is consistent with the data types facets</li>
				<li>Target-Disease Evidence page: somatic mutations from the Cancer Gene Census are presented by mutation type with the number of samples recorded for each mutation type. The cellular mechanism of the mutation (dominant/recessive) is also given based on the Cancer Gene Census annotation</li>
				<li>Target-Disease Evidence: for known drug evidence, the clinical trial phase status is presented</li>
		    </ul>
		</div>
	</div>
	<!-- release 1.0 -->
	<div class="col-md-12">
		<div>
			<h2>Release 1.0 (release 2015-12-08)</h2>
			<h3>Status</h3>
			<p>This is the first public version of the CTTV platorm</p> 
			<p>The CTTV platform integrates over 2M target-disease associations covering 7,359 diseases and 24,309 targets from the following public resources:</p>
			<ul>
				<li>9,600 ontology classes from <a href="http://www.ebi.ac.uk/efo/">the Experimental Factor Ontology</a> (EFO version 2.66)</li>
				<li>9,741 genetic evidence from <a href="http://www.uniprot.org/">UniProt</a>, a comprehensive, high-quality and freely accessible resource of protein sequence and functional information</li>
				<li>19,084 genetic evidence from <a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a>, an open-access database of all types of genetic variation data from human</li>
				<li>25,876 genetic evidence from <a href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>, the NHGRI-EBI Catalog of published genome-wide association studies</li>
				<li>20,830 evidence from <a href="https://www.ebi.ac.uk/chembl/">ChEMBL</a>, a database that contains compound bioactivity data against drug targets</li>
				<li>377,360 evidence from <a href="https://www.ebi.ac.uk/gxa/home">Expression Atlas</a> that provides information on gene expression patterns under different biological conditions</li>
				<li>9,390 affected pathway evidence from <a href="http://www.reactome.org/">Reactome</a>, a free, open-source, curated and peer reviewed pathway database</li>
				<li>12,079 evidence from the <a href="http://cancer.sanger.ac.uk/census/">COSMIC cancer Gene Census</a>, a catalogue of genes for which mutations have been causally implicated in cancer</li>
				<li>459 somatic evidence from <a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a></li>
				<li>440,601 evidence from <a href="https://www.sanger.ac.uk/resources/databases/phenodigm">PhenoDigm</a>, a database of animal model matches ranked by their phenotypic similarity to human disorder</li>
				<li>3,327,695 scientific literature evidence from co-occurence mining in <a href="https://europepmc.org/">Europe PMC</a></li>
			</ul>
			<p>This release includes the following features described in details in the <a href="/about">documentation</a>:</p>
			<ul>
				<li>Search box: you can retrieve targets of interest by entering gene symbols, gene names, gene synonyms, Ensembl identifiers and UniProt accessions</li>
				<li>Search box: you can also search by disease name provided the disease label and its synonyms are part of the Experimental Factor Ontology (EFO)</li>
				<li>Search box: you can also search by phenotype provided the phenotype label and its synonyms are included in the Experimental Factor Ontology (EFO)</li>
				<li>Search results: results of a search are presented by order of relevance based on the matching of symbols, names and synonyms. Matched terms are highlighted</li>
				<li>Search filters: search results can be filtered by target and/or disease. Number of matching targets and number of matching diseases are shown</li>
				<li>Auto-completion box: search terms are auto-completed. the auto-completion presents - up to 3 - matched target and disease terms</li>
				<li>Auto-completion box: EFO disease ontology classification is presented under the disease term and clickable. A disease hyperlink links to the disease association page</li>
				<li>Auto-completion box: the hyperlink of an auto-completed target links to a Target Association page for this target</li>
				<li>Target Association page: provides an overview of all diseases associated to a target. 3 views are available: a 'bubble view', a 'tabular view' and a 'tree' view</li>
				<li>Target Association page bubble view: groups disease by therapeutic area into bubbles. Because of the EFO ontology, a disease can belong to multiple therapeutic area</li>
				<li>Target Association page bubble view: you can click on any bubble to get the association score between a specific disease and a target. This score is broken down by data types and represented as a flower</li>
				<li>Target Association page bubble view: you can go to a detailed target-disease Evidence page by following the view details link when clicking on a bubble</li>
				<li>Target Association page tabular view: lists all diseases associated to a particular gene ordered by association score. This score is broken down by data types</li>
				<li>Target Association page tree view: represents diseases associated to a specific target at different classification levels of the EFO ontology. Nodes in the tree can be collapsed or expanded</li>
				<li>Target Association page filters: filter by datatypes and therapeutic areas</li>
				<li>Disease Association page: provides a tabular view of all potential targets associated to a disease ordered by association score. This score is broken down by data types</li>
				<li>Target-Disease Evidence page: provides details on the evidence found to associate a particular target to a particular disease (from EFO). The association score flower represents the relative contribution of each data type when scoring a target to disease association</li>
				<li>Target-Disease Evidence: genetic associations evidence from eVA, UniProt and the GWAS catalog are presented. A genome browser is embedded to display mutations</li>
				<li>Target-Disease Evidence: somatic mutations evidence from the cancer gene census COSMIC resource and from eVA are presented</li>
				<li>Target-Disease Evidence: known drug evidence from ChEMBL are presented</li>
				<li>Target-Disease Evidence: RNA expression evidence from the Expression Atlas are presented</li>
				<li>Target-Disease Evidence: target-disease co-occurence evidence mined from Europe PMC are presented</li> 
				<li>Target-Disease Evidence: disrupted pathways evidence from Reactome are presented</li>
				<li>Target-Disease Evidence: mouse models of disease from PhenoDigm are presented</li>
				<li>Target Profile page: summarises all information known about a particular target including synonyms, UniProt and Ensembl information, GO terms, baseline protein and RNA expression and list of publications from EUROPE PubMed Central</li>
				<li>Disease Profile page: provides a description of a disease, its synonyms and its classification in the Experimental Factor Ontology</li>
		    </ul>			
			<p>The CTTV platform also integrates third-party visualisations components:</p>
			<ul>
				<li>The human baseline expression widget from the EMBL-EBI <a href="https://www.ebi.ac.uk/gxa/home">Expression Atlas</a></li>
				<li>The protein features widget from <a href="http://www.uniprot.org/">UniProt</a></li>
				<li>The pathway navigation widget from <a href="http://www.reactome.org/">Reactome</a></li>
				<li>The protein structure visualisation is extracted from <a href="https://www.ebi.ac.uk/pdbe/">PDBe</a></li>
			</ul>
		</div>
	</div>	
</div>
