<view-title>Release Notes</view-title>

<h1>Center for Therapeutic Target Validation Release Notes</h1>

<div class="row">

	<div class="col-md-12">

		<div>
			<h2>Release 0.6</h2>
			<h3>Status</h3>
			<p>This is a planned updated version of the CTTV platform that includes several improvements and fixes in the web application. The following items were fixed and/or added:</p>
			<ul>
				<li>association page: Consistent facets in both pages (for targets and diseases)</li>
				<li>association page for targets: Consistent <i>therapeutic area</i> navigation accross the three views (bubbles, tables and tree)</li>
				<li>association page for target: Better transitions in bubbles view when filters are applied</li>
				<li>target profile page: Dropped dependency on external widget for Uniprot literature display and added number of citations per publication</li>
				<li>target profile page: Updated the expression atlas widget</li>
				<li>general: better support for private mirror by allowing reauthentication when session expires</li>
			</ul>
			<p>In addition, the following items were fixed:</p>
			<ul>
				<li>association page for targets: Fixed inconsistent number of associated diseases in the different views / filters</li>
				<li>target profile page: Fixed links to reactome in pathways section</li>
				<li>target profile page: Fixed bug that prevented the genome browser to render because of malformed SNPs</li>
				<li>target profile page: Fixed experiment links in the expression atlas widget</li>
			</ul>

			<h2>Release 0.5 (release 2015-07-10)</h2>
			<h3>Status</h3>
			<p>This is a planned updated version of the CTTV platorm that includes several improvements and fixes in the web application. The following items were fixed and/or added:</p>
			<ul>
				<li>search: exact match on synonym search</li>
				<li>documentation: a new <a href="/about">About</a> page with general information about the CTTV web platform. A short description has also being added to the <a href="/">main page</a> of the site</li>
				<li>target associations page: The datatypes filter is now open by default</li>
				<li>disease associations page: targets associated to the children terms of the EFO disease are now returned</li>
				<li>disease associations page: improved facet behaviour - data are now updated based on the filter selection</li>
				<li>disease associations page: improved column sorting on the heatmap</li>
				<li>disease profile page: the disease classification is now shown as a graph view with links to navigate to the parents and children disease terms</li>
				<li>target profile page: new 'Target Family' section added showing the phylogenetic tree of the target</li>
				<li>target profile page: added a short description on some of the sections ("Variants, isoforms and genomic context" and "Protein Baseline Expression" and "Target Family")</li>
            </ul>
			<p>In addition, the following items were fixed:</p>
			<ul>
			    <li>Genomic browser: fixed the links from SNPs to bibliographic references in the tooltips</li>
				<li>Target profile page: fixed formatting of small real numbers in the tooltips</li>
			</ul>

			<h2>Release 0.4 (release 2015-06-15)</h2>
			<h3>Status</h3>
			<p>This is a planned updated version of the CTTV platform. The following items were fixed and/or added:</p>
			<ul>
				<li>target profile page: <i>Variants, isoforms and genomic context</i> section improved to display the location of SNPs associated to diseases. This is now referred as the <i>BioJS variation browser widget</i></li>
				<li>target profile page: fixed Internet Explorer 9 issues with Expression Atlas widget</li>
				<li>target profile page: fixed Internet Explorer 9 issues with the Bibliography widget</li>
				<li>target association page: 'data types' facet filters added to select sources of evidence. Selecting the filters refreshes the heatmap</li>
				<li>target association page: 'Reactome pathways' facet filters added to filter the list of targets in the heatmap</li>
				<li>disease associations page: heatmap view color palette improved to distinguish empty cells from cells with lower score</li>
				<li>disease associations page: ‘Tree’ view was modified to show the top-level disease categories at a fixed position</li>
				<li>disease associations page: 'Tree' view now shows labels for internal nodes in the tree</li>
				<li>disease associations page: tooltip added to display the full disease name when mouse over a bubble or a tree node</li>
				<li>disease associations page: clicking on a bubble or a tree node will open a pop-up with additional direct links to the disease page or the associations page for the disease</li>
				<li>gene to disease association page: data tables have been reformatted to display more relevant information</li>
				<li>gene to disease association page: GWAS data table - scientific notations for p-values</li>
				<li>gene to disease association page: RNA expression data table - percentile rank added to indicate the relative expression level of that gene in comparison to all other genes for this experiment</li>
				<li>gene to disease association page: BioJS variation browser widget added to the genetics evidence part of the page. The variation view is enabled when clicking on the 'Browser' tab below the 'Genetic associations' title</li>
				<li>heatmaps: Fixed Internet Explorer 9 issue with the heatmaps legend</li>
			</ul>

			<p>Data Changes, including updated and refined several of the data sources, particularly known drugs and their targets:</p>
			<ul>
				<li>major update of ChEMBL disease associations (June 2015)</li>
				<li>new version of GWAS Catalog disease associations (May 2015)</li>
				<li>new version of the Expression Atlas disease associations (June 2015)</li>
				<li>new version of EVA disease associations (May 2015)</li>
				<li>new version of EFO with description for over 150 Orphanet rare disease terms</li>
				<li>Gene assigment to haplotypes fixed in mouse models</li>
			</ul>

			<h2>Release 0.3 (release 2015-05-19)</h3>
				<h3>Status</h3>
				<p>This is a maintenance update of the CTTV platform. The following items were fixed and/or added:</p>
				<ul>
					<li>search box: higher coverage of synonyms from HGNC and UniProt</li>
					<li>disease associations page: intermediate disease terms are now shown in the bubble view</li>
					<li>disease associations page: the table view has been improved in the form of a heatmap to make a better use of the space on the page. Scores are represented by a color and the flower has been removed</li>
					<li>target page: transcript information display added to show alternative transcripts of a gene with location of SNPs</li>
					<li>target page: tissue expression display of the Human Protein Atlas baseline expression</li>
					<li>target page: Gene Ontology terms display improved</li>
					<li>target page: Reactome pathway section added</li>
					<li>target page: Drug section added</li>
					<li>disease page: display of immediate EFO children of a disease in the disease path section</li>
				</ul>

				<p>Data Changes:</p>
				<ul>
					<li>Expression profiles of human genes on the mRNA and protein level from the Human Protein Atlas</li>
					<li>Ensembl genes mapped to non-reference assembly were discarded to avoid duplication of symbols</li>
				</ul>
				<p>In addition, the following items were fixed:</p>
				<ul>
					<li>search results: duplicated genes</li>
				</ul>
				<p>Known bugs:</p>
				<ul>
					<li>tabular export to pdf can generate overlapping rows</li>
				</ul>

				<h2>Release 0.2 (release 2015-04-27)</h2>
				<h3>Status</h3>
				<p>This is a planned updated version of the CTTV platform. The following items were fixed and/or added:</p>
				<ul>
					<li>search box: better support of synonyms, including protein names</li>
					<li>search box: better handling of partial word search and non alphanumerical chars</li>
					<li>search results: filter checkboxes for disease and target checked by default</li>
					<li>association score: simplified score calculation easier to understand</li>
					<li>association score: animal model score changed reflecting the score provided in PhenoDigm</li>
					<li>disease associations page: data types facet filters are now clickable to refine the evidence types that are shown</li>
					<li>disease associations page: new ‘Tree’ view that allows visualisation of evidence across the therapeutic areas in a tree representing the relationships of diseases</li>
					<li>disease associations page: data types facet filters are also visible on the Table and Tree view</li>
					<li>disease associations data types filters: the number of associated diseases is indicated next to each data type</li>
					<li>disease associations data types filters: clicking/unclicking on a data type filter updates the Bubble, Table or Tree representation accordingly</li>
					<li>disease associations bubble view: support for multiple EFO paths</li>
					<li>gene to disease association page: export (to excel, pdf) feature added for every data type</li>
					<li>disease page: display multiple EFO paths</li>
					<li>URI changed to be resource oriented for disease, target and association pages</li>
					<li>The word 'gene' was replaced by the preferred word 'target' across the pages</li>
					<li>generally improved consistency across the evidence pages and refined the external links</li>
					<li>added secure authentication to the system so that you will be able to access by a secure, authenticated connection</li>
					<li>added documentation including Release Notes and answers to Frequently Asked Questions </li>
				</ul>

				<p>Data Changes, including updated and refined several of the data sources, particularly in genetics:</p>
				<ul>
					<li>new version of EVA (version 0.9) disease associations</li>
					<li>new version of GWAS Catalog disease associations (April 2015)</li>
					<li>new version of mouse models of human disease (phenoDigm) predictions based on IMPC (version 3.0) and MGI data (February 2015)</li>
					<li>new version of EFO with increased coverage of diseases including over 1,000 disease terms from EVA mapped to EFO</li>
					<li>new version of Ensembl (release 79) for genomic location and identifiers</li>
					<li>updated version of HGNC approved gene symbols</li>
				</ul>
				<p>In addition, the following items were fixed:</p>
				<ul>
					<li>target to disease association page: mouse models are now listed properly</li>
					<li>search for gene symbol with identical prefix (e.g. BRD) works</li>
				</ul>

				<h2>Release 0.1 (release 2015-03-02)</h2>
				<h3>Status</h3>
				<p>This version is the first release of the CTTV platform. The following major features were added:</p>
				<ul>
					<li>search box: you can retrieve gene of interest by entering gene symbol, gene name and gene synonym</li>
					<li>search box: you can also search by disease name provided the disease label and its synonyms are part of the Experimental Factor Ontology (EFO)</li>
					<li>search results: results of a search are presented by order of relevance based on the matching of symbol, names and synonyms. Matched terms are highlighted</li>
					<li>search filters: search results can be filtered by gene and/or disease. Number of matching genes and number of matching diseases are shown</li>
					<li>auto-completion box: search terms are auto-completed. the auto-completion presents - up to 3 - matched gene and disease terms </li>
					<li>auto-completion box: disease ontology classification is presented under the disease term and clickable. A disease hyperlink links to the disease association page</li>
					<li>auto-completion box: the hyperlink of an auto-completed gene links to a gene associations page for this gene</li>
					<li>disease associations page: provides an overview of all diseases associated to a target. 2 views are available: a 'bubble view' and a 'tabular view'</li>
					<li>disease associations bubble view: groups disease by therapeutic area into bubbles. Because of the EFO ontology, a disease can belong to multiple therapeutic area</li>
					<li>disease associations bubble view: you can zoom in a particular therapeutic area by clicking on the corresponding therapeutic area filter</li>
					<li>disease associations bubble view: you can zoom out of a particular therapeutic area by clicking on the corresponding therapeutic area filter again</li>
					<li>disease associations bubble view: you can click on a bubble to get the association score between a specific disease and a target. This score is broken down by data types and represented as a flower</li>
					<li>disease associations bubble view: you can go to a detailed gene-disease page by following the view details link when clicking on a bubble</li>
					<li>disease associations tabular view: lists all diseases associated to a particular gene ordered by association score. This score is broken down by data types and represented as a flower</li>
					<li>association score flower: represents the relative contribution of each data type when scoring a gene to disease association</li>
					<li>gene to disease association page: provides details on the evidence found to associate a particular gene to a particular disease (from EFO)</li>
					<li>gene to disease association page: genetic associations from eVA, UniProt and the GWAS catalog are presented</li>
					<li>gene to disease association page: somatic mutations from the cancer gene census COSMIC resource are presented</li>
					<li>gene to disease association page: known drugs from ChEMBL are presented</li>
					<li>gene to disease association page: RNA expression information from the Expression Atlas are presented</li>
					<li>gene to disease association page: disrupted pathways from Reactome are presented</li>
					<li>gene to disease association page: mouse models from PhenoDigm are presented</li>
					<li>target profile page: summarises all information known about a particular target including synonyms, UniProt and Ensembl information, GO terms, baseline expression and list of publications from EUROPE PubMed Central</li>
					<li>disease profile page: provides a description of a disease, its synonyms and its classification in the Experimental Factor Ontology</li>
					<li>target association page: provides a table view of all potential targets associated to a disease as a table</li>
					<li>target associations tabular view: lists all targets associated to a particular disease ordered by association score. This score is broken down by data types and represented as a flower</li></li>
				</ul>
				<p>The CTTV platform integrates multiple source of information from the following public resources:</p>
				<ul>
					<li><a href="http://www.uniprot.org/">UniProt</a>: a comprehensive, high-quality and freely accessible resource of protein sequence and functional information</li>
					<li><a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a>: an open-access database of all types of genetic variation data from human</li>
					<li><a href="https://www.ebi.ac.uk/gwas/">The GWAS Catalog</a>: the NHGRI-EBI Catalog of published genome-wide association studies</li>
					<li><a href="https://www.ebi.ac.uk/chembl/">ChEMBL</a>: a database that contains compound bioactivity data against drug targets</li>
					<li><a href="https://www.ebi.ac.uk/gxa/home">The Expression Atlas</a> provides information on gene expression patterns under different biological conditions such as a gene knock out, a plant treated with a compound, or in a particular organism part or cell</li>
					<li><a href="http://cancer.sanger.ac.uk/census/">The COSMIC cancer Gene Census</a>: a catalogue of genes for which mutations have been causally implicated in cancer</li>
					<li><a href="https://www.sanger.ac.uk/resources/databases/phenodigm">PhenoDigm</a>: a database of animal model matches ranked by their phenotypic similarity to human disorder</li>
				</ul>

			</div>
		</div>
	</div>
