<view-title>Release Notes</view-title>

<h1>Center for Therapeutic Target Validation Release Notes</h1>

<div class="row">
	<div class="col-md-12">
		<div>
			<h2>Release 1.0 (release 2015-12-07)</h2>
			<h3>Status</h3>
			<p>This is the first public version of the CTTV platorm that includes the following features described in details in the <a href="">documentation</a>:</p>
			<ul>
				<li>search box: you can retrieve targets of interest by entering gene symbols, gene names, gene synonyms, Ensembl identifiers and UniProt accessions</li>
				<li>search box: you can also search by disease name provided the disease label and its synonyms are part of the Experimental Factor Ontology (EFO)</li>
				<li>search box: you can also search by phenotype provided the phenotype label and its synonyms are included in the Experimental Factor Ontology (EFO)</li>
				<li>search results: results of a search are presented by order of relevance based on the matching of symbols, names and synonyms. Matched terms are highlighted</li>
				<li>search filters: search results can be filtered by target and/or disease. Number of matching targets and number of matching diseases are shown</li>
				<li>auto-completion box: search terms are auto-completed. the auto-completion presents - up to 3 - matched target and disease terms</li>
				<li>auto-completion box: EFO disease ontology classification is presented under the disease term and clickable. A disease hyperlink links to the disease association page</li>
				<li>auto-completion box: the hyperlink of an auto-completed target links to a Target Association page for this target</li>
				<li>Target Association page: provides an overview of all diseases associated to a target. 3 views are available: a 'bubble view', a 'tabular view' and a 'tree' view</li>
				<li>Target Association page bubble view: groups disease by therapeutic area into bubbles. Because of the EFO ontology, a disease can belong to multiple therapeutic area</li>
				<li>Target Association page bubble view: you can click on any bubble to get the association score between a specific disease and a target. This score is broken down by data types and represented as a flower</li>
				<li>Target Association page bubble view: you can go to a detailed target-disease Evidence page by following the view details link when clicking on a bubble</li>
				<li>Target Association page tabular view: lists all diseases associated to a particular gene ordered by association score. This score is broken down by data types</li>
				<li>Target Association page tree view: represents diseases associated to a specific target at different classification levels of the EFO ontology. Nodes in the tree can be collapsed or expanded</li>
				<li>Target Association page filters: filter by datatypes and therapeutic areas</li>
				<li>Disease Association page: provides a tabular view of all potential targets associated to a disease ordered by association score. This score is broken down by data types</li>
				<li>Evidence page: provides details on the evidence found to associate a particular target to a particular disease (from EFO). The association score flower represents the relative contribution of each data type when scoring a target to disease association</li>
				<li>Evidence page: genetic associations evidence from eVA, UniProt and the GWAS catalog are presented. A genome browser is embedded to display mutations</li>
				<li>Evidence page: somatic mutations evidence from the cancer gene census COSMIC resource and from eVA are presented</li>
				<li>Evidence page: known drug evidence from ChEMBL are presented</li>
				<li>Evidence page: RNA expression evidence from the Expression Atlas are presented</li>
				<li>Evidence page: target-disease co-occurence evidence mined from Europe PMC are presented</li> 
				<li>Evidence page: disrupted pathways evidence from Reactome are presented</li>
				<li>Evidence page: mouse models of disease from PhenoDigm are presented</li>
				<li>target profile page: summarises all information known about a particular target including synonyms, UniProt and Ensembl information, GO terms, baseline protein and RNA expression and list of publications from EUROPE PubMed Central</li>
				<li>disease profile page: provides a description of a disease, its synonyms and its classification in the Experimental Factor Ontology</li>
		    </ul>
			<p>The CTTV platform integrates multiple source of information from the following public resources:</p>
			<ul>
				<li>The Experimental Factor Ontology - EFO (2.66) with description for over x ontology classes (November 2015)</li>
				<li><a href="http://www.uniprot.org/">UniProt</a>: a comprehensive, high-quality and freely accessible resource of protein sequence and functional information</li>
				<li><a href="http://www.ebi.ac.uk/eva/">European Variation Archive (EVA)</a>: an open-access database of all types of genetic variation data from human</li>
				<li><a href="https://www.ebi.ac.uk/gwas/">The GWAS Catalog</a>: the NHGRI-EBI Catalog of published genome-wide association studies</li>
				<li><a href="https://www.ebi.ac.uk/chembl/">ChEMBL</a>: a database that contains compound bioactivity data against drug targets</li>
				<li>377,360 evidence from <a href="https://www.ebi.ac.uk/gxa/home">the Expression Atlas</a> provides information on gene expression patterns under different biological conditions</li>
				<li>9,390 affected pathway evidence from <a href="http://www.reactome.org/"></a>Reactome</a></li>
				<li><a href="http://cancer.sanger.ac.uk/census/">The COSMIC cancer Gene Census</a>: a catalogue of genes for which mutations have been causally implicated in cancer</li>
				<li><a href="https://www.sanger.ac.uk/resources/databases/phenodigm">PhenoDigm</a>: a database of animal model matches ranked by their phenotypic similarity to human disorder</li>
				<li><a href="https://europepmc.org/">Europe PMC</a>:integration of 3,327,695 literature mining evidence from co-occurence mining</li>
			</ul>
			<p>The CTTV platform integrates third-party widgets developed and maintained by external public resources:</p>
			<ul>
				<li>The human baseline expression viewer cfrom the EMBL-EBI Expression Atlas</li>
				<li>The protein features viewer from UniProt</li>
				<li>The pathway navigation widget from Reactome</li>
				<li>The protein structure visualisation is extracted from ePDB</li>
			</ul>
		</div>
	</div>
</div>
